pageserver/src/tenant/tasks.rs

```diff
@@ -407,9 +407,16 @@ async fn gc_loop(tenant: Arc<Tenant>, cancel: CancellationToken) {
                         error_run_count += 1;
                         let wait_duration = Duration::from_secs_f64(wait_duration);
 
-                        error!(
-                        "Gc failed {error_run_count} times, retrying in {wait_duration:?}: {e:?}",
-                    );
+                        if matches!(e, crate::tenant::GcError::TimelineCancelled) {
+                            // Timeline was cancelled during gc. We might either be in an event
+                            // that affects the entire tenant (tenant deletion, pageserver shutdown),
+                            // or in one that affects the timeline only (timeline deletion).
+                            // Therefore, don't exit the loop.
+                            info!("Gc failed {error_run_count} times, retrying in {wait_duration:?}: {e:?}");
+                        } else {
+                            error!("Gc failed {error_run_count} times, retrying in {wait_duration:?}: {e:?}");
+                        }
+
```

Comment:
Just realized, should a single timeline cancellation fail an entire run of gc? I don't see the point.

Reply:
I've been wondering about that as well, and decided to handle it at the top level, because I wasn't sure how a tenant cancellation works, it might cancel the timelines one by one, so something that's actually a tenant cancellation might show up as timeline cancellations here, so if we continued to the next timeline, we would find that one to be cancelled as well, etc etc.

So I thought better not touch the tenant for the next couple of seconds, shrug. An erroneous gc run does not mean we only run one at the next regularly scheduled point, it will be retried a couple seconds later.